### PR TITLE
MGDOBR-750: Manager is not able to process batch of new Bridges

### PR DESCRIPTION
**Pull request by manstis**
[https://issues.redhat.com/browse/MGDOBR-750](https://issues.redhat.com/browse/MGDOBR-750)

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

**Comment by manstis** [comment]
@robpblake @sutaakar This is a draft proposal to better handle resource provisioning.

This uses Quartz (in memory for now, on this PR) using the basics of Rob's [proposal](https://issues.redhat.com/browse/MGDOBR-750?focusedCommentId=20552682&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-20552682).

I'd like your opinions before I spend a lot of time refactoring the tests.

**Comment by manstis** [comment]
I've tested this locally and created/deleted a bunch of Bridges/Processors/Managed Connectors and it's great :+1: 

IDK if it's _easy_ to use this with our integration/load tests to see if if provides an improvement.

**Comment by manstis** [comment]
@sutaakar Can you please advise how to run the integration tests locally?

I've followed the [instructions](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/tree/main/integration-tests) but only manage to see all tests fail with the following reason:
```
[ERROR] SendToBridge Action tests - Send Cloud Event to SendToBridge with bridgeId defined  Time elapsed: 0.27 s  <<< ERROR!
java.lang.IllegalStateException: Cannot parse object because no supported Content-Type was specified in response. Content-Type was 'text/html'.
```

**Comment by sutaakar** [comment]
@manstis I ran the tests using `./run-local-tests.sh -t wip` (and tagging `SendToBridge Action tests` as `@wip`) and it passed successfully.
What steps did you do to run the E2E tests?

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java [hunk 1]
```diff
@@ -44,22 +55,40 @@
     @ConfigProperty(name = "event-bridge.resources.workers.timeout-seconds")
     int timeoutSeconds;
 
+    @ConfigProperty(name = "event-bridge.resources.workers.frequency")
```
[the remark this answers is not in the document]
Comment (sutaakar) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java:58]: + to adjust this property in other places

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java [hunk 2]
```diff
@@ -48,7 +52,23 @@ protected PanacheRepositoryBase<ConnectorEntity, String> getDao() {
...
+        if (Objects.isNull(connectorEntity)) {
+            //Work has been scheduled but cannot be found. Something (horribly) wrong has happened.
+            throw new IllegalStateException(String.format("Connector for Processor with id '%s' cannot be found in the database.", processorId));
+        }
+
+        return connectorEntity.getId();
+    }
+
+    @Override
+    @ActivateRequestContext
```
Comment (sutaakar) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java:70]: What is this annotation used for?

File: manager/src/main/resources/db/migration/0.1/V0.1.16__add_quartz_scheduler.sql [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+CREATE SEQUENCE hibernate_sequence start 1 increment 1;
+
+CREATE TABLE TASKS
+(
+    id        int8 NOT NULL,
+    createdAt TIMESTAMP,
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE QRTZ_JOB_DETAILS
```
Comment (sutaakar) [inline manager/src/main/resources/db/migration/0.1/V0.1.16__add_quartz_scheduler.sql:24]: Are these SQL commands taken from some specific external location?
Are they supposed to change in some future Quartz version or will remain the same?

Comment (manstis) [reply manager/src/main/resources/db/migration/0.1/V0.1.16__add_quartz_scheduler.sql:24]: It was lifted from [Quarkus](https://github.com/quarkusio/quarkus-quickstarts/blob/main/quartz-quickstart/src/main/resources/db/migration/V2.0.0__QuarkusQuartzTasks.sql) (linked from [here](https://quarkus.io/guides/quartz)) which seems to follow the [Quartz](https://github.com/quartz-scheduler/quartz/blob/master/quartz-core/src/main/resources/org/quartz/impl/jdbcjobstore/tables_postgres.sql) DDL.

I have however now added the Quartz DDL (as I notice Quarkus added an additional table).

As to whether they change.. they could and it is a risk; but I'd assume it's the same for any consumer of Quartz with a database backed `JobStore` (Quartz's storage abstraction) and something we'd need to cover. Our integration tests would fail if a Quartz/Quarkus version change broke the DDL. Our Unit Tests use an in-memory `JobStore` and hence are not sensitive to DDL issues.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java (hunk 2 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java:70]: This is nice... 

`ConnectorWorker` uses `ConnectorsApiClient` which is `@RequestScoped` however Quartz is running `ConnectorWorker` on a thread that does not have a request scope active. This annotation ensures CDI activates a request scope to prevent use of `ConnectorsApiClient` from failing. I went for this solution than changing the scope of `ConenctorsApiClient` as that could have had further reaching consequences.

**Comment by manstis** [comment]
@robpblake @sutaakar This PR is ready for review/merge.

It addresses the "thundering herd" issue @robpblake mentions on the JIRA by re-scheduling each resource 30s after each initial request. There however remains the issue that a `Uni` (to grant access to a RHOAS topic) can complete after it has been timed-out [elsewhere](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/43e13e3c557476ac2ba8a08ee02295c0ccc8689d/manager/src/main/java/com/redhat/service/smartevents/manager/RhoasServiceImpl.java#L49). 

However since the time-out is 15s and our re-try mechanism will run for 90s before failing the likelihood of an issue is now much smaller (but not completely eliminated). The "thundering herd" issue is (probably) the bigger issue as it puts a lot of stress on RHOAS so perhaps we should pragmatically consider this PR as a partial solution and another JIRA for the Uni/time-out issue?

File: manager/src/main/resources/db/migration/0.1/V0.1.16__add_quartz_scheduler.sql (hunk 3 above)
Comment (sutaakar) [reply manager/src/main/resources/db/migration/0.1/V0.1.16__add_quartz_scheduler.sql:15]: ok, thanks for info

**Review by sutaakar — APPROVED** [review]
Looks good to me, works as expected

**Comment by robpblake** [comment]
@manstis This is on my list of things to get to. Just ack'ing your review request.

**Comment by robpblake** [comment]
@manstis Apologies for the delay in getting to this. It's on my list for first thing tomorrow. A couple of questions/comments I do have:

- What impact does moving to a per-item reschedule interval have on our failure rate versus the single work-loop? If we create `30` Bridge instances in one go, with this approach do they all enter the `READY` state versus a % of them being `FAILED` on the single work loop?

 - Can you please test what happens if we have multiple instances of Fleet Manager trying to process the queue. In particular I'm interested to understand 
   -  Is work correctly processed if the Fleet Manager is shutdown and then restarted at a period of time later?
   -  Is the clustered handling in Quartz working as we expect it to with the queue being processed across multiple Fleet Manager instances, or "confirming our hope that Quartz lets us remove our custom `reconnectDroppedWorkers` implementation.

cheers,
Rob

**Comment by afalhambra** [comment]
> I was hoping to be able to use the Performance Tests https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/915/ to make this task easier. @sutaakar @afalhambra do you know whether this would make a suitable load test for Bridge/Processor creation?

@manstis I think this would be a good candidate for running a load test scenario as part of the performance test suite. I can start working on that one. wdyt @sutaakar?

**Comment by sutaakar** [comment]
@afalhambra I can cover that test.

@manstis it can be tested, just thinking what results do we actually expect....
Is a Bridge creation failure considered as expected outcome or a failure outcome? What metrics are we interested in?
- Speed of Bridge/Processor creation (number of seconds until READY state)?
- Reliability (number of failed Bridges/Processors)?
Should we define a threshold for these metrics to fail the test? Or just produce a report to be analyzed manually (possibly automated in future)?

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/WorkManagerImpl.java [hunk 4]
```diff
@@ -2,124 +2,87 @@
...
+        try {
+            TriggerKey key = new TriggerKey(id, MANAGED_RESOURCES_GROUP);
+            if (Objects.nonNull(quartz.getTrigger(key))) {
+                LOGGER.info(String.format("Work already exists for '%s'. Skipping.", id));
+            }
 
-    @SuppressWarnings("unused")
-    @Transactional
-    @Scheduled(every = "5m", concurrentExecution = Scheduled.ConcurrentExecution.SKIP)
-    void reconnectDroppedWorkers() {
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/WorkManagerImpl.java:120]: @manstis nice to see `reconnectDroppedWorkers` disappearing from our codebase :-)

File: manager/src/test/resources/application.properties [hunk 5]
```diff
@@ -20,6 +20,8 @@ event-bridge.kafka.sasl.mechanism=PLAIN
...
 event-bridge.rhoas.mgmt-api.host=http://test-management-api.redhat.com
 
-# Workers require scheduling however we only enable it for the applicable tests.
-# See Quarkus Test Profile WorkerSchedulerProfile
-quarkus.scheduler.enabled=false
+# Quartz Scheduler configuration for tests
+event-bridge.resources.workers.frequency-seconds=1
+quarkus.quartz.start-mode=forced
+quarkus.quartz.clustered=false
+quarkus.quartz.store-type=ram
```
Comment (robpblake) [inline manager/src/test/resources/application.properties:27]: @manstis Is there any reason why we can't or shouldn't use the JDBC store in our tests?

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java [hunk 6]
```diff
@@ -154,4 +180,29 @@ protected boolean isTimeoutExceeded(Work work, ManagedResource managedResource)
...
+            TriggerKey existingTriggerKey = existingTrigger.getKey();
+            Trigger newTrigger = TriggerBuilder.newTrigger()
+                    .forJob(existingTrigger.getJobKey())
+                    .withIdentity(existingTriggerKey)
+                    .usingJobData(existingTrigger.getJobDataMap())
+                    .startAt(new Date(System.currentTimeMillis() + frequencySeconds * 1000))
+                    .build();
+            quartz.rescheduleJob(existingTriggerKey, newTrigger);
+        } catch (SchedulerException e) {
+            LOGGER.error(String.format("Unable to reschedule Job for '%s", context.getTrigger().getJobDataMap().getString(STATE_FIELD_ID)), e);
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java:204]: @manstis We need to consider how to handle this failure because otherwise I think the associated resource (Bridge or Processor) simply ends-up stuck in the current `status`. Is my understanding correct?

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/WorkManagerImpl.java [hunk 7]
```diff
@@ -2,124 +2,87 @@
...
+                    .withIdentity(key)
+                    .usingJobData(STATE_FIELD_ID, id)
+                    .usingJobData(STATE_FIELD_ATTEMPTS, String.valueOf(0))
+                    .usingJobData(STATE_FIELD_SUBMITTED_AT, ZonedDateTime.now(ZoneOffset.UTC).toString())
+                    .startNow()
+                    .build();
+            quartz.scheduleJob(trigger);
+
+        } catch (SchedulerException e) {
+            LOGGER.info(String.format("Unable to schedule work for '%s'", id), e);
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/WorkManagerImpl.java:85]: @manstis I think we need to throw this `Exception`. If we don't then the associated resource will be forever stuck in the `ACCEPTED` status. By throwing the `Exception`, we can rollback the DB transaction commit which will prevent the resource from being created (or updated).

**Comment by robpblake** [comment]
> 

@manstis Thanks for the investigations into processing across multiple instances of the Fleet Manager. That's great to hear 👍 and I think gives us the confidence to remove our `reconnectDroppedWorkers` logic.

Before we move forwards with this, I would like us to try and get some data that it improves our situation with handling multiple `Bridge` / `Processor` creation requests at the same time. From the original issue, @sutaakar seemed to suggest that at about `30` concurrent bridges, we run into some significant difficulty getting them into the `READY` status.

@manstis @sutaakar @afalhambra Ideally if we can use the load testing work here that would be great. I think the first metric for us to focus on is reliability, e.g can we get data that show (I'm making up numbers):

- Before this change, at `30` concurrent bridges, 75% of those bridges enter the `FAILED` status after bridge creation request timeout
- After this change, all `30` bridges are in the `READY` status after the bridge creation timeout.

cheers,

Rob

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/WorkManagerImpl.java [hunk 8]
```diff
@@ -2,124 +2,87 @@
...
-        return workDAO.findByWorkerId(workerId);
-    }
-
-    @Transactional(Transactional.TxType.NEVER)
-    protected void processWorkQueue(List<Work> work) {
-        work.forEach(this::fireEvent);
-    }
+    // Schedule the first execution of a Job.
+    // We cannot use Quartz scheduler to invoke the job repeatedly as we store mutable state within the
+    // Trigger definition and Quartz does not support updating the Trigger JobDataMap after the Trigger
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/WorkManagerImpl.java:63]: This comment is still valid. 

`AbstractWorker` handles the re-scheduling after updating the `Trigger`'s `JobDataMap`. I cannot schedule the `Job` to be repeating with Quartz and remove it when complete as that approach does not support updating the `JobDataMap` to record the number of (failed) attempts.
- I can schedule once, update data and re-schedule.
- I cannot schedule repeating and update the job data on each invocation.

File: manager/src/test/resources/application.properties (hunk 5 above)
Comment (manstis) [reply manager/src/test/resources/application.properties:27]: Speed... otherwise no reason at all.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/WorkManagerImpl.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/WorkManagerImpl.java:63]: @manstis Thanks for the explanation.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/quartz/QuartzWorkManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Override
+    public boolean exists(ManagedResource managedResource) {
+        Work work = Work.forResource(managedResource);
+        TriggerKey key = new TriggerKey(work.getManagedResourceId(), work.getType());
+        try {
+            return quartz.checkExists(key);
+        } catch (SchedulerException e) {
+            String message = "Failed to check if a Job exists for resource of type '" + work.getType() + "' and id '" + work.getManagedResourceId() + "'";
+            throw new IllegalStateException(message, e);
```
Comment (danielezonca) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/quartz/QuartzWorkManagerImpl.java:89]: Platform exception? (Same comment for the other similar code)

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ProcessorWorker.java [hunk 10]
```diff
@@ -64,7 +63,17 @@ public Processor createDependencies(Work work, Processor processor) {
 
         // If we have to deploy a Managed Connector, delegate to the ConnectorWorker.
         // The Processor will be provisioned by the Shard when it is in ACCEPTED state *and* Connectors are READY (or null).
-        return delegate(work, processor);
+        String processorId = work.getManagedResourceId();
+        ConnectorEntity connectorEntity = connectorsDAO.findByProcessorId(processorId);
+        ConnectorEntity updatedConnectorEntity = connectorWorker.createDependencies(work, connectorEntity);
+        processor.setDependencyStatus(updatedConnectorEntity.getStatus());
+
+        // If the Connector failed we should mark the Processor as failed too
```
Comment (danielezonca) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ProcessorWorker.java:71]: Is this code new or moved from somewhere else?

**Review by danielezonca — COMMENTED** [review]
I had a look to the PR and added some comments, my approval is not mandatory so feel free to proceed when you have enough reviews :)

Btw most of the comments are minor

File: manager/src/test/java/com/redhat/service/smartevents/manager/api/user/ProcessorAPITest.java [hunk 11]
```diff
@@ -705,15 +706,20 @@ public void updateProcessorWithTemplate() {
                 bridge.getId(),
                 new ProcessorRequest("myProcessor", filters, "template", TestUtils.createKafkaAction()));
 
-        ProcessorResponse processor = TestUtils.getProcessor(bridge.getId(), createResponse.as(ProcessorResponse.class).getId()).as(ProcessorResponse.class);
-        setProcessorAsReady(processor.getId());
+        // We have to wait until the existing Work to provision the Processor is complete.
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/test/java/com/redhat/service/smartevents/manager/api/user/ProcessorAPITest.java:709]: Yes, @robpblake request. We now schedule work `now() + frequency` which for tests is `now() + 1s`

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/quartz/QuartzWorkManagerImpl.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    @Inject
+    org.quartz.Scheduler quartz;
+
+    private JobDetail workJob;
+
+    @PostConstruct
+    public void init() {
+        try {
+            workJob = JobBuilder.newJob(WorkJob.class).storeDurably().build();
+            quartz.addJob(workJob, false);
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/quartz/QuartzWorkManagerImpl.java:56]: No. We define one `Job` and then multiple `Trigger`'s to execute the `Job` with different parameters (defined in `Work` i.e. `ManagedResource.id`, submitted date, number of attempts etc). Since we re-use the same `Job` we simply declare it to Quartz once in start-up.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/quartz/WorkJob.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+import com.redhat.service.smartevents.manager.models.Bridge;
+import com.redhat.service.smartevents.manager.models.Processor;
+import com.redhat.service.smartevents.manager.models.Work;
+import com.redhat.service.smartevents.manager.workers.Worker;
+import com.redhat.service.smartevents.manager.workers.resources.BridgeWorker;
+import com.redhat.service.smartevents.manager.workers.resources.ProcessorWorker;
+
+import static com.redhat.service.smartevents.manager.workers.quartz.QuartzWorkConvertor.convertFromJobData;
+
+// Single Job implementation that basically represents what we have now in the Work table queue.
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/quartz/WorkJob.java:17]: Done.

**Comment by manstis** [comment]
@robpblake @danielezonca all concerns have been addressed/refactored as appropriate.

@robpblake @sutaakar has provided some promising metrics too.

Can we progress this PR to approval and merge?

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java [hunk 14]
```diff
@@ -153,6 +169,7 @@ private ConnectorEntity deployConnector(ConnectorEntity connectorEntity) {
     }
 
     @Override
+    @ActivateRequestContext
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java:172]: @manstis Could you add a comment here just to explain why the `@ActivateRequestContext` is required (and should not be removed).

cheers,
Rob

File: manager/src/test/java/com/redhat/service/smartevents/manager/utils/DatabaseManagerUtils.java [hunk 15]
```diff
@@ -62,7 +58,6 @@ public void cleanUpAndInitWithDefaultShard() {
     @Transactional
     public void cleanUp() {
         // Clean up
-        deleteAllWork();
```
Comment (robpblake) [inline manager/src/test/java/com/redhat/service/smartevents/manager/utils/DatabaseManagerUtils.java:65]: @manstis Do we need to do anything to clear up the `Quartz` tables at all?

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java [hunk 16]
```diff
@@ -48,6 +49,21 @@ protected PanacheRepositoryBase<ConnectorEntity, String> getDao() {
     }
 
     @Override
+    protected String getId(Work work) {
```
Comment (robpblake) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/ConnectorWorker.java:52]: @manstis I don't know if this is actually ever called. It looks like `ProcessorWorker` calls the create/delete dependencies method directly on `ConnectorWorker`?

File: manager/src/test/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorkerTest.java [hunk 17]
```diff
@@ -111,8 +113,13 @@ void testProvisionWorkWithKnownResource(ManagedResourceStatus status,
         Bridge refreshed = worker.handleWork(work);
 
         assertThat(refreshed.getDependencyStatus()).isEqualTo(dependencyStatusWhenComplete);
-        assertThat(workManager.exists(work)).isNotEqualTo(isWorkComplete);
         verify(rhoasServiceMock).createTopicAndGrantAccessFor(TEST_TOPIC_NAME, RhoasTopicAccessType.CONSUMER_AND_PRODUCER);
+        verify(workManager, times(isWorkComplete ? 0 : 1)).reschedule(any());
+    }
+
+    @Transactional
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/test/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorkerTest.java:120]: No, it's not.. I suspect a rogue copy and paste at sometime. Deleted.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java [hunk 18]
```diff
@@ -49,17 +49,15 @@
...
     public T handleWork(Work work) {
-        T managedResource = load(work);
+        String id = getId(work);
+        T managedResource = load(id);
         if (Objects.isNull(managedResource)) {
             //Work has been scheduled but cannot be found. Something (horribly) wrong has happened.
-            workManager.complete(work);
-            String message = String.format("Resource of type '%s' with id '%s' no longer exists in the database.", work.getType(), work.getManagedResourceId());
-            throw new IllegalStateException(message);
+            throw new IllegalStateException(String.format("Resource with id '%s' no longer exists in the database.", id));
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java:56]: Reverted.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java [hunk 19]
```diff
@@ -72,44 +70,33 @@ public T handleWork(Work work) {
...
-                        e);
+                LOGGER.info(String.format("Failed to delete dependencies for '%s'.", id), e);
                 // Something has gone wrong. We need to retry.
-                workManager.recordAttempt(work);
+                workManager.rescheduleAfterFailure(work);
             }
         }
 
-        if (complete) {
-            workManager.complete(work);
```
[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java:104]: I'll check tomorrow and comment here.

[the remark this answers is not in the document]
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/AbstractWorker.java:104]: OK, so, I ran this PR and created some things. I observed Quartz writing to its tables. I also observed Quartz cleaning up its tables when the `Work` was complete. The end result was that after everything had ran Quartz's tables were empty. I therefore conclude no explicit application (RHOSE) clean-up code is needed.

File: manager/src/test/java/com/redhat/service/smartevents/manager/utils/DatabaseManagerUtils.java (hunk 15 above)
Comment (manstis) [reply manager/src/test/java/com/redhat/service/smartevents/manager/utils/DatabaseManagerUtils.java:65]: See https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/920#discussion_r922037832.

tl;dr; No.

**Comment by manstis** [comment]
@robpblake @danielezonca This is ready for your approvals :+1:

**Comment by manstis** [comment]
@danielezonca @robpblake Could you please try to find some time to review.

**Comment by manstis** [comment]
Merged.

@danielezonca [commented](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/920#pullrequestreview-1036292939) "...my approval is not mandatory so feel free to proceed when you have enough reviews".
